public/app/features/explore/ExploreRunQueryButton.tsx

```diff
@@ -0,0 +1,137 @@
+import { uniq } from 'lodash';
+import React, { useState } from 'react';
+import { ConnectedProps, connect } from 'react-redux';
+
+import { DataSourceApi } from '@grafana/data';
+import { config, getDataSourceSrv, reportInteraction } from '@grafana/runtime';
+import { DataQuery } from '@grafana/schema';
+import { Button, Dropdown, Menu, ToolbarButton } from '@grafana/ui';
+import { t } from '@grafana/ui/src/utils/i18n';
+import { useSelector } from 'app/types';
+
+import { changeDatasource } from './state/datasource';
+import { setQueries } from './state/query';
+import { isSplit, selectExploreDSMaps, selectPanesEntries } from './state/selectors';
+
+const mapDispatchToProps = {
+  setQueries,
+  changeDatasource,
+};
+
+const connector = connect(undefined, mapDispatchToProps);
+
+interface ExploreRunQueryButtonProps {
+  queries: DataQuery[];
+  rootDatasourceUid?: string;
+}
+
+export type Props = ConnectedProps<typeof connector> & ExploreRunQueryButtonProps;
+
+export function ExploreRunQueryButton({ rootDatasourceUid, queries, changeDatasource, setQueries }: Props) {
+  const [openRunQueryButton, setOpenRunQueryButton] = useState(false);
+  const isPaneSplit = useSelector(isSplit);
+  const exploreActiveDS = useSelector(selectExploreDSMaps);
+  const panesEntries = useSelector(selectPanesEntries);
+
+  const isDifferentDatasource = (uid: string, exploreId: string) =>
+    !exploreActiveDS.dsToExplore.find((di) => di.datasource.uid === uid)?.exploreIds.includes(exploreId);
+
+  // exploreId on where the query will be ran, and the datasource ID for the item's DS
+  const runQueryText = (exploreId: string, dsUid?: string) => {
+    // if the datasource or exploreID is undefined, it will be disabled, but give it default query button text
+    return dsUid !== undefined && exploreId !== undefined && isDifferentDatasource(dsUid, exploreId)
+      ? {
+          fallbackText: 'Switch data source and run query',
+          translation: t('explore.run-query.switch-datasource-button', 'Switch data source and run query'),
+        }
+      : {
+          fallbackText: 'Run query',
+          translation: t('explore.run-query.run-query-button', 'Run query'),
+        };
+  };
+
+  const runQuery = async (exploreId: string) => {
+    const differentDataSource = isDifferentDatasource(rootDatasourceUid!, exploreId);
+    setQueries(exploreId, queries);
+
+    reportInteraction('grafana_explore_query_history_run', {
+      queryHistoryEnabled: config.queryHistoryEnabled,
+      differentDataSource,
+    });
+  };
+
+  const validateDatasources = async (datasourceUids: Array<string | undefined>): Promise<Promise<boolean>> => {
```

Comment:
So it looks like a bug -  in mixed we always shows the button enabled. To me it feels like it should be tackled in a separate PR to make it easier to review/track. I would prefer to have features and bugs committed separately (e.g. in case feature needs to be reverted/rolled back - we don't want to roll back the fix :D)

Not sure if we need to pull the API of the data source. Maybe getting settings `dataSourceSrv.getInstanceSettings` is enough (it's synchronous so it's simpler). And as you mentioned we already have some sort of validation in `DatasourceInfo` component so might be better to do it outside run button. The solution may depend on UX feedback as well, and we may try out different things so might be better to isolate it to a seprate isse. wdyt?

Reply:
https://github.com/grafana/grafana/issues/88001